New segfault in build during inference

I see a new segfault 100% of the time when building RPM nightlies using LLVM 3.7.1 (including the three Julia patches). It has been introduced over the last two days.

I can do some debugging if you tell me what you need to know.

```
cd /builddir/build/BUILD/julia/base && /builddir/build/BUILD/julia/build/usr/bin/julia -C x86-64 --output-ji /builddir/build/BUILD/julia/build/usr/lib64/julia/inference0.ji -f coreimg.jl
/bin/sh: line 1: 14588 Segmentation fault      (core dumped) /builddir/build/BUILD/julia/build/usr/bin/julia -C x86-64 --output-ji /builddir/build/BUILD/julia/build/usr/lib64/julia/inference0.ji -f coreimg.jl
Makefile:214: recipe for target '/builddir/build/BUILD/julia/build/usr/lib64/julia/inference0.ji' failed
```
https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/fedora-23-x86_64/00159825-julia/build.log.gz

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -1075,22 +1075,26 @@ extern "C" void jl_generate_fptr(jl_lambda_info_t *li)
         JL_SIGATOMIC_BEGIN();
         #ifdef USE_MCJIT
         if (imaging_mode) {
-            // Copy the function out of the shadow module
-            Module *m = new Module("julia", jl_LLVMContext);
-            jl_setup_module(m);
-            FunctionMover mover(m, shadow_module);
-            mover.CloneFunction((Function*)li->functionObjects.functionObject);
-            if (li->functionObjects.specFunctionObject != NULL)
-                mover.CloneFunction((Function*)li->functionObjects.specFunctionObject);
-            if (li->functionObjects.cFunctionList != NULL) {
-                size_t i;
-                cFunctionList_t *list = (cFunctionList_t*)li->functionObjects.cFunctionList;
-                for (i = 0; i < list->len; i++) {
-                    list->data()[i].f = mover.CloneFunction(list->data()[i].f);
-                }
-            }
-            jl_finalize_module(m);
+            // see if it has been emitted already (as part of compiling something else)
             li->fptr = (jl_fptr_t)jl_ExecutionEngine->getFunctionAddress(((Function*)li->functionObjects.functionObject)->getName());
+            if (li->fptr == NULL) {
+                // Copy the function out of the shadow module
+                Module *m = new Module("julia", jl_LLVMContext);
+                jl_setup_module(m);
+                FunctionMover mover(m, shadow_module);
+                mover.CloneFunction((Function*)li->functionObjects.functionObject);
+                if (li->functionObjects.specFunctionObject != NULL)
+                    mover.CloneFunction((Function*)li->functionObjects.specFunctionObject);
+                if (li->functionObjects.cFunctionList != NULL) {
+                    size_t i;
+                    cFunctionList_t *list = (cFunctionList_t*)li->functionObjects.cFunctionList;
+                    for (i = 0; i < list->len; i++) {
+                        list->data()[i].f = mover.CloneFunction(list->data()[i].f);
+                    }
+                }
+                jl_finalize_module(m);
+                li->fptr = (jl_fptr_t)jl_ExecutionEngine->getFunctionAddress(((Function*)li->functionObjects.functionObject)->getName());
+            }
         }
         else {
             li->fptr = (jl_fptr_t)getAddressForOrCompileFunction((Function*)li->functionObjects.functionObject);
@@ -3689,14 +3693,12 @@ static void finalize_gc_frame(Function *F)
 
 static void finalize_gc_frame(Module *m)
 {
-#if defined(USE_ORCJIT)
+#if defined(USE_MCJIT) || defined(USE_ORCJIT)
     for (auto &F : m->functions()) {
         if (F.isDeclaration())
             continue;
         finalize_gc_frame(&F);
     }
-#endif
-#if defined(USE_MCJIT) || defined(USE_ORCJIT)
 #ifndef JULIA_ENABLE_THREADING
     m->getFunction("jl_get_ptls_states")->eraseFromParent();
 #endif

